src/Bundle/CoreBundle/Field/Types/ReferenceFieldType.php

```diff
@@ -268,13 +268,13 @@ function validateData(FieldableField $field, $data, ExecutionContextInterface $c
         }
 
         // Only validate available data.
-        if (empty($data)) {
+        if (empty($data) && !$field->getSettings()->not_empty) {
```

Comment:
A test case for this change is missing. This is the important test case for this PR, because it fixes the bug.

Reply:
I've updated the reference field config test to include the `not_empty` setting, and added a new content controller test that covers saving a `not_empty` reference field.